nim_status/client.nim

```diff
@@ -362,6 +394,17 @@ proc toWalletAccount(account: accounts.Account): WalletAccount {.used.} =
   let name = if account.name.isNone: "" else: account.name.get
   WalletAccount(address: account.address, name: name)
 
+proc getCustomTokens*(self: StatusObject): CustomTokensResult =
+  if not self.isLoggedIn:
+    return CustomTokensResult.err "Not logged in. Must be logged in to get " &
+      "custom tokens."
+  try:
+    let tokens = self.userDb.getCustomTokens()
+    return CustomTokensResult.ok tokens
+  except Exception as e:
```

Comment:
`Exception` should never be caught, it will cause issues with the distinction between `Defect` and `CatchableError` and doesn't work predictably - normally, more specific errors should be used instead, see - https://status-im.github.io/nim-style-guide/errors.exceptions.html

Reply:
In the case where unknown errors are filtering through, you can catch `CatchableError` instead, though we strive to avoid that

Reply:
Thanks for the tips @arnetheduck. After reviewing the style guide, there are quite a lot of changes we need to make as we are catching `Exception` in a lot of places. 

In this particular case, there are two branches of exceptions that we'd like to catch for this db call:
1. Any errors coming from the sql execution in the db (nim-sqlcipher lib)
2. An exception raised in the `userDb` proc (this needs to be updated to raise a custom error, but for this example let's call it `DbNotOpenError`)

What's the best way to proceed here? Would it be to catch both the `DbNotOpenError` and also to catch `CatchableError` which may come from any sql execution errors? Or should we try to modify nim-sqlcipher in a way that only raises custom errors?

Reply:
So there are two scenarios in general that are worth considering, I think - one is when dealing with user-supplied code, when you're making a "framework" - in that scenario, you can't anticipate what errors can happen but the nature of error handling tends to be limited to mitigating damage (closing sockets, releasing resources etc) and passing the exception on to someone more competent to handle it - this is not that common however (ie chronos and task_runner qualify but most other projects not) 

the other scenario is when you're creating abstraction layers - in this latter scenario, every non-custom error is an abstraction leak and therefore an architectural problem - ie when you're creating a "status client", the consumer of that client should not have to deal with "db error" from the db layer because there's nothing reasonable that they can do with this information beyond showing it to the user (which makes for terrible UX) - in particular, they are not in a position to import the concrete type of the error and reason about it, because that is an internal implementation detail of "the client". 

In such cases, the recommendation is that each layer handles _all_ exceptions originating from layers below it and either takes action (retries, repairs, reconnects, whatever) or reframes the error in terms of the "current" abstraction layer - "filesystem error" becomes "database error" becomes "cannot persist message" in the os, database and "message client" sections of the code respectively, each layer not only bundling functionality at a higher abstraction level, but also the possible errors that come out of there.

Finally, a layer such as `userDb` depends a little bit: if it adds its own "error hierarchy", one way to deal with it is to create a `UserDbError` in the general "Db" hierarchy of exceptions that has fields for the "inner" error that was raised from the `userDb` call, allowing consumers of these exceptions to catch `UserDbError` separately, _then_ drilling into it to find the "deeper" error if they care.

Important here is the design of any callback as well - callbacks that don't raise exceptions are easiest to reason about, but otherwise, you should document the callback itself with raises annotations (see https://status-im.github.io/nim-style-guide/language.proctypes.html)

> make as we are catching Exception in a lot of places.

I'm not sure if you're following it already, but the first and most important change is to start adding `{.push raises: [Defect].}` at the top of every module to enable _some_ error tracing in the language - the next step is to work on getting rid of `Exception` catches - part of this process is to _also_ sort out implicit Exception effects. There are more steps after that , but this is a start.

This is work that is best done early on in the project because inevitably it will need doing or the client will leak and crash and you won't be able to address it without rearchitecting the code completely to introduce the required error handling - as the other Nim projects we have have learned the hard way.

Reply:
Thanks so much for the writeup @arnetheduck. @michaelsbradleyjr and I have read through it thoroughly a few times and it has provided quite a lot of insight. We discussed it further and it resulted in a small [exception handling proposal](https://discord.com/channels/624205794384281629/720285913858703391/869740387089125376), which is based on our limited understanding. The first part of this is in draft (https://github.com/status-im/nim-status/pull/246) and we would love to get your feedback on it if possible 🙏